### PR TITLE
Add group membership extension view function

### LINKED ISSUE #86: GroupMembership Extension view function
We need to provide the functionality to allow a group owner/manager wants to extend a group membership.

We need a view function that:
- takes a GroupMembership pk as an argument.
- checks that the accessing user is an owner/manager of the group in question (or a superadmin).
- on GET request - displays the user's information and the current expiry date as well as a form allowing them to specify the length of an extension.
- on POST request - updates the expiry date of the GroupMembership based on the form data and redirects to `group_members_view`.

This is a typical get/post pattern for a view with a form. The `user_search` view function is a simple example to pattern off of.

### PR DESCRIPTION
**Pull request by SaranjeetKaur**
# Description

This PR adds a new group membership extension view function using the group membership pk as an argument.

To do:
- [x] check that the accessing user is an owner/manager of the group in question (or a superadmin).
- [x] on GET request - display the user's information and the current expiry date as well as a form allowing them to specify the length of an extension.
- [x] on POST request - updates the expiry date of the GroupMembership based on the form data and redirects to `group_members_view`.
- [x] add unit test

Fixes #86 

## Type of change

- [ ] Documentation (non-breaking change that adds or improves the documentation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking, back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Key checklist

- [x] All tests pass (eg. `python -m pytest`)
- [ ] The documentation builds and looks OK (eg. `python -m sphinx -b html docs docs/build`)
- [x] Pre-commit hooks run successfully (eg. `pre-commit run --all-files`)

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added or an issue has been opened to tackle that in the future. (Indicate issue here: # (issue))

### CODE REVIEW

File: imperial_coldfront_plugin/templates/imperial_coldfront_plugin/extend_membership.html [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{% extends "common/base.html" %}
+{% block content %}
+  <p>Your current membership is due to expire on {{ expiry_date }}.</p>
+  Would you like to extend it?
```
Comment (cc-a) [inline imperial_coldfront_plugin/templates/imperial_coldfront_plugin/extend_membership.html:4]: Currently looks like this is addressed to the user that is being extended rather than the user doing the extension.

File: imperial_coldfront_plugin/views.py [hunk 2]
```diff
@@ -366,3 +366,61 @@ def remove_group_manager(
...
+    # extension.
+
+    if request.method == "GET":
+        return render(
+            request,
+            "imperial_coldfront_plugin/extend_membership.html",
+            {
+                "group_membership": group_membership,
+                "group": group,
+            },
```
Comment (cc-a) [inline imperial_coldfront_plugin/views.py:407]: Not currently passing the form in here.

File: imperial_coldfront_plugin/views.py [hunk 3]
```diff
@@ -366,3 +366,61 @@ def remove_group_manager(
...
+        )
+
+    # on POST request - updates the expiry date of the GroupMembership based on the
+    # form data and redirects to group_members_view.
+
+    if request.method == "POST":
+        form = GroupMembershipForm(request.POST)
```
Comment (cc-a) [inline imperial_coldfront_plugin/views.py:414]: You'll need a dedicated form for this I think.

**Review by cc-a — COMMENTED** [review]
Definitely looking on the right track. I've put some comments, though you were probably aware of these already.

**Comment by SaranjeetKaur** [comment]
might add some more unit tests for the view

File: tests/test_views.py [hunk 4]
```diff
@@ -570,3 +572,53 @@ def test_successful_manager_removal(self, pi, pi_client, pi_group, mailoutbox):
...
+        response = client.get(self._get_url())
+        assert response.status_code == HTTPStatus.FORBIDDEN
+        assert response.content == b"Permission denied"
+
+    def test_group_owner(self, pi_client, pi_group):
+        """Test that the group owner can extend a group membership."""
+        group_membership = pi_group.groupmembership_set.first()
+        response = pi_client.get(self._get_url(group_membership.pk))
+        assert response.status_code == HTTPStatus.OK
+        assert isinstance(response.context["form"], GroupMembershipExtendForm)
```
Comment (cc-a) [inline tests/test_views.py:600]: Need to also check that the group membership is included in the context as this is currently missing which causes info to be missing when the page is rendered.

**Review by cc-a — CHANGES_REQUESTED** [review]
Thanks @SaranjeetKaur. Looking very nice, good catch on the manager extending their own membership corner case. In addition to my comment I think this should have some tests for the manager related behaviours i.e. managers can access and extend memberships except their own.

**Review by cc-a — APPROVED** [review]
Great. Thanks for this.
